promql/engine.go

```diff
@@ -379,6 +415,31 @@ func (ng *Engine) exec(ctx context.Context, q *query) (Value, storage.Warnings,
 	ctx, cancel := context.WithTimeout(ctx, ng.timeout)
 	q.cancel = cancel
 
+	defer func() {
+		ng.queryLoggerLock.RLock()
+		if l := ng.queryLogger; l != nil {
+			f := []interface{}{"query", q.q}
+			switch eq := q.Statement().(type) {
+			case *EvalStmt:
+				f = append(f,
+					"start", formatDate(eq.Start),
+					"end", formatDate(eq.End),
+					"step", eq.Interval.String(),
+				)
+			case testStmt:
+				f = append(f, "test", "yes")
+			}
+			if v := ctx.Value(queryCtx("ip")); v != nil {
```

Comment:
This looks nice, but similarly to the previous comments, it would be maybe nice to allow users of `promql` package to add their own fields e.g duration, tenant ID, additional statistics etc. 

We can do that by: 
A) Allow adding and fetching context value for queryLogFields
B) Handle query log outside of the engine. There is nothing specific/private that we don't have access to outside e.g in webHandler. It can also handle log file reopen.

There is nothing blocking `promql` non-Prometheus users to reimplement similar query log thing with their own fields outside  of this as well, so not a blocker (:

Reply:
The stats are private.

Reply:
hm.. `q.Stats().ToMap()` is public?

Reply:
Indeed. However recording rules etc also need to be taken into account. I do not import config anymore.